src/parser/symtab.ts

```diff
@@ -0,0 +1,1738 @@
+        before analyzing names, because class bindings aren't
+        visible in methods.  For other blocks, they are initialized
+        after names are analyzed.
+        */
+
+        /* TODO(jhylton): Package these dicts in a struct so that we
+        can write reasonable helper functions?
+        */
+        const newglobal = new globalThis.Set<string>();
+        const newfree = new globalThis.Set<string>();
+        const newbound = new globalThis.Set<string>();
+
+        /* Class namespace has no effect on names visible in
+        nested functions, so populate the global and bound
+        sets to be passed to child blocks before analyzing
+        this one.
+        */
+        if (ste.blockType == ClassBlock) {
+            /* Pass down known globals */
+            append(newglobal, global);
+
+            /* Pass down previously bound symbols */
+            if (bound) {
+                append(newbound, bound);
+            }
+        }
+
+        for (const [name, flags] of Object.entries(ste.symbols)) {
+            this.analyze_name(ste, scopes, name, flags, bound, local, free, global);
+        }
+
+        /* Populate global and bound sets to be passed to children. */
+        if (ste.blockType !== ClassBlock) {
+            /* Add function locals to bound set */
+            if (ste.blockType === FunctionBlock) {
+                append(newbound, local);
+            }
+            /* Pass down previously bound symbols */
+            if (bound) {
+                append(newbound, bound);
+            }
+            /* Pass down known globals */
+            append(newglobal, global);
+        } else {
+            /* Special-case __class__ */
+            newbound.add("__class__");
+        }
+
+        /* Recursively call analyze_child_block() on each child block.
+        newbound, newglobal now contain the names visible in
+        nested blocks.  The free variables in the children will
+        be collected in allfree.
+        */
+        const allfree = new globalThis.Set<string>();
+        for (const entry of ste.children) {
+            this.analyze_child_block(entry, newbound, newfree, newglobal, allfree);
+            if (entry.hasFree || entry.childHasFree) {
+                ste.childHasFree = true;
+            }
+        }
+
+        append(newfree, allfree);
+
+        /* Check if any local variables must be converted to cell variables */
+        if (ste.blockType === FunctionBlock) {
+            this.analyze_cells(scopes, newfree);
+        } else if (ste.blockType === ClassBlock) {
+            ste.drop_class_free(newfree);
+        }
+
+        /* Records the results of the analysis in the symbol table entry */
+        ste.update_symbols(ste.symbols, scopes, bound, newfree, ste.blockType === ClassBlock);
+
+        append(free, newfree);
+    }
+}
+
+export function BuildSymbolTable(mod: mod, filename: string, future: never): SymbolTable {
```

Comment:
Maybe worth splitting this module up? I almost feel like each class should be its own module.
And then functions like this that are for interaction should also be in a separate module.
(Deno uses `mod.ts` as the common way to export things from a common file)